### PR TITLE
Wide wave track

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #5624
Resolves: #5198
Resolves: #5159
Resolves: #6081

Growing pull request in progress will make a real wide WaveTrack structure, eliminating most need for the old
track list link flag, except during deserialization.

There will not yet be a real wide WaveClip, unified with the WideClip and WaveTrack::Interval notions.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

QA:
- [x] Saving, closing and re-opening project with both mono and stereo tracks works, also if these tracks have effect stacks
- [x] Doing this across Audacity versions also works
- [x] If you have any old project that has a stereo track whose channels have different sample rates and/or sample formats, please try to open it, and see that the error messages are still as expected.
- [x] Moving clips from track to track, both mono and stereo, one or several at a time, still works.
- [x] The problem reported in #5264 hasn't reappeared.
- [x] Split stereo to mono -- including examples with nontrivial envelopes
- [x] Join mono to stereo, also mono tracks with different settings (e.g. color, level, mute/solo)
- [x] Use some Nyquist generator (e.g. Pluck) on both mono and stereo tracks
- [x] Delete clips from mono and stereo tracks
- [x] Mix Stereo to Mono in the Tracks menu
- [x] Generate noise, mono and stereo
- [x] Click and drag to zoom vertical rulers in waveform and spectrogram
- [x] Vertical ruler context menu items
- [x] Dragging of MIDI data too
- [x] Track > Duplicate for mono and stereo; also cut, copy, paste
- [x] Naming of new clips is unchanged when splitting, recording, etc.
- [x] Dragging clips (left, right, up, or down) is correctly prevented from making overlaps
- [x] Benchmark tool
- [x] Sorting of clips by name and by time
- [x] opening of 2.4.2 (or earlier) projects with mono and stereo tracks
- [x] Copy-paste with difference of stretch ratio
- [x] Swapping stereo channels, including with envelopes
- [x] Test some destructive effects
- [x] Rearranging sequence of mono and stereo tracks, dragging one up or down
- [x] Both R and Shift + R recording, mono and stereo
- [x] Simply making a new empty stereo track
- [x] Long undo/redo message, in the History window, after Mix and Render, correctly describes the new track as mono or stereo
- [x] Force-quit Audacity, then restart exercise crash recovery, either recovering or discarding
- [x] Open files from earlier versions (either .aup or .aup3) with misaligned clips
- [x] Reverse effect

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
Rebased unchanged

**Comment by Paul-Licameli** [comment]
Another rebase, with no change except one easy conflict resolution with 6110c4f5e73d90c367ba606c7a47bd8352c0f7c3

**Comment by Paul-Licameli** [comment]
Rebased to updated dependency branch

**Comment by Paul-Licameli** [comment]
Rebased, squashed some fix-ups, added a new first commit in the sequence

**Review by saintmatthieu — COMMENTED** [review]
Reviewed commits up to [Define and use WaveChannelView::GetFirst and FindFirst](https://github.com/audacity/audacity/pull/5965/commits/539cd3c862caa97cae1df90274a29a935ab2edca) exclusively - to be continued

File: libraries/lib-music-information-retrieval/tests/benchmarking-dataset [hunk 1]
```diff
@@ -0,0 +1 @@
+../../../../tempo-detection-benchmarking-dataset-2
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-music-information-retrieval/tests/benchmarking-dataset:1]: Mistakenly added

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 2]
```diff
@@ -3619,24 +3619,31 @@ sampleFormat WaveTrack::WidestEffectiveFormat() const
 bool WaveTrack::HasTrivialEnvelope() const
 {
    auto pTrack = this;
-   if (GetOwner())
+   if (pTrack->GetOwner())
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3762]: I was thinking of moving it to a nonmember function in wavetrackutilities

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 3]
```diff
@@ -3619,24 +3619,31 @@ sampleFormat WaveTrack::WidestEffectiveFormat() const
...
-      [](const auto &pClip){ return pClip->GetEnvelope()->IsTrivial(); });
+      [](const auto &pClip){ return pClip->GetEnvelope().IsTrivial(); });
 }
 
 void WaveTrack::GetEnvelopeValues(
    double* buffer, size_t bufferLen, double t0, bool backwards) const
 {
    auto pTrack = this;
-   if (GetOwner())
+   if (pTrack->GetOwner())
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3783]: Ditto

**Comment by saintmatthieu** [comment]
Looking at `class WideChannelGroupInterval : public ChannelGroupInterval`, I'm wondering what need there is to distinguish the two. A channel _group_ is necessarily wide, isn't it?

**Review by saintmatthieu — DISMISSED** [review]
Reviewed commits up to [Replace WaveClip with Interval in two more functions for time shift](https://github.com/audacity/audacity/pull/5965/commits/b93a44be31b3cfde0e79402d6108bc95c6a9826f), inclusive.
A crash bug was found, and I would like that the suggestion about the `const` method overload is applied.

**Comment by Paul-Licameli** [comment]
Rebased onto updated wave-clip-refactoring.  Made the newly merged in handle classes compile.  Removed the extraneous file.

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 4]
```diff
@@ -3089,11 +3087,11 @@ bool WaveTrack::Append(size_t iChannel,
    // TODO wide wave tracks -- there will be only one clip, and its `Append`
    // (or an overload) must take iChannel
    auto pTrack = this;
-   if (GetOwner() && iChannel == 1)
-      pTrack = *TrackList::Channels(this).rbegin();
    constSamplePtr buffers[]{ buffer };
-   return pTrack->RightmostOrNewClip()
-      ->Append(buffers, format, len, stride, effectiveFormat);
+   auto iter =  RightmostOrNewClip()->Channels().begin();
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:3091]: What steps exactly reproduce the crash?

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 5]
```diff
@@ -1495,18 +1495,34 @@ ChannelGroup &WaveTrack::DoGetChannelGroup() const
 TrackListHolder WaveTrack::Clone(bool backup) const
 {
    assert(IsLeader());
-   auto result = TrackList::Temporary(nullptr);
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:1436]: Avoids violating some `IsLeader()` assertions.

[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:1436]: Put wave tracks into the temporary Tracklist objects before doing other things with them, while they are still not yet wide tracks.

File: libraries/lib-wave-track/WaveTrack.h [hunk 6]
```diff
@@ -222,6 +222,34 @@ class WAVE_TRACK_API WaveChannel
...
    CallbackReturn Publish(const WaveTrackMessage &message);
 
+   AudioGraph::ChannelType GetChannelType() const override;
+
+   size_t NChannels() const override;
+
+   //! Takes gain and pan into account
+   float GetChannelGain(int channel) const override;
+
+   //! This fails if any clip overlapping the range has non-unit stretch ratio!
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.h:232]: I don't understand -- delete the comment?  Or complete it, how?

Or you can do that as you will in master.

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 7]
```diff
@@ -623,6 +622,14 @@ auto WaveTrack::Interval::GetCutLines(WaveTrack &track)
    return result;
 }
 
+auto WaveTrack::Interval::GetCutLines(const WaveTrack &track) const
+   -> IntervalConstHolders
+{
+   auto results =
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:637]: I don't really agree.  It is not too hard by inspection of the non-const function to see that it doesn't change any data structure directly, but just exposes non-const pointers.  This pattern of the const calling the non-const with a cast is a common enough idiom.  The total lines of code is less as I did it.

File: libraries/lib-note-track/NoteTrack.h [hunk 8]
```diff
@@ -190,10 +190,15 @@ class NOTE_TRACK_API NoteTrack final
...
-      using WideChannelGroupInterval::WideChannelGroupInterval;
+   struct Interval final : WideChannelGroupInterval {
+      explicit Interval(const NoteTrack &track);
       ~Interval() override;
       std::shared_ptr<ChannelInterval> DoGetChannel(size_t iChannel) override;
+      double Start() const override;
+      double End() const override;
+   private:
+      //! @invariant not null
+      const std::shared_ptr<const NoteTrack> mpTrack;
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-note-track/NoteTrack.h:201]: I like it as I wrote it.  Less is assumed about lifetimes of things this way.  Maybe intervals would survive.

I had reasons to rewrite WaveTrack::Interval to hold shared pointers, not references.  But I admit, there, the pointers are to WaveClip objects, not the track.

File: libraries/lib-wave-track/WaveTrack.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp]: Never mind, I understand the problem, which happened not to crash on the debug macOS build, but that was an "accident."

File: src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp [hunk 9]
```diff
@@ -129,15 +129,20 @@ class WaveTrackShifter final : public TrackShifter {
       // TODO wide wave tracks -- simplify when clips are really wide
       auto pRight = mpTrack->ChannelGroup::GetChannel<WaveTrack>(1);
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp]: ok

File: libraries/lib-wave-track/WaveTrack.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (saintmatthieu) [reply libraries/lib-wave-track/WaveTrack.cpp:637]: Do the const-cast in the direction I propose and no inspection is needed anymore but that of the const-cast of the return type. Then the real implementation may grow and there still won't be a need for inspection. Your arguments not to apply this suggestion aren't solid.

**Review by saintmatthieu — COMMENTED** [review]
Review complete up to "Fewer direct uses of WaveTrack::mClips..." inclusively.

**Comment by Paul-Licameli** [comment]
Rebased unchaged to updated base branch

File: libraries/lib-wave-track/WaveTrack.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:637]: No.

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 10]
```diff
@@ -1789,79 +1728,91 @@ WaveTrack::Holder WaveTrack::EmptyCopy(
...
 {
    assert(!GetOwner());
+   mRightChannel.reset();
+   EraseChannelAttachments(1);
 
-   // Make new track
+   // Make temporary new mono track
    auto result = Duplicate();
-   auto &right = static_cast<WaveTrack&>(**result->begin());
+   auto &newTrack = static_cast<WaveTrack&>(**result->begin());
```
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp:1775]: not used

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 11]
```diff
@@ -4310,41 +4247,42 @@ bool WaveChannel::ClipsAreUnique() const
...
+   auto iter = pOwner->Find(this);
+   assert(this == *iter);
+   ++iter;
+   assert(iter != pOwner->end()); // pre
+   auto pRight = dynamic_cast<WaveTrack*>(*iter);
+   assert(pRight && pRight->NChannels() == 1); // pre
+
+   //! Refuse if clips are not well aligned.
+   if (!AreAligned(this->SortedClipArray(), pRight->SortedClipArray()))
       return;
```
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp:4272]: Shouldn't the modification `this->SetLinkType(LinkType::None);` be undone here?

File: libraries/lib-wave-track/CMakeLists.txt [hunk 12]
```diff
@@ -35,8 +35,6 @@ set( SOURCES
    WaveTrackSink.h
    WaveTrackUtilities.cpp
    WaveTrackUtilities.h
-   WideClip.cpp
```
Comment (saintmatthieu) [inline libraries/lib-wave-track/CMakeLists.txt:38]: Great to see these gone!
I was expecting this commit to justify the previous one, a67b7aded083eea4a3fbdfb455e95801e49a60f6, but does it, actually?

**Review by saintmatthieu — COMMENTED** [review]
Reviewed up to fa016813abed83226788e55566239ed485761759, inclusively.

File: libraries/lib-track/ChannelAttachments.cpp [hunk 13]
```diff
@@ -126,29 +126,42 @@ bool ChannelAttachmentsBase::HandleXMLAttribute(
...
+         pAttachment->Reparent(parent, ii);
 }
 
-void ChannelAttachmentsBase::Erase(size_t index)
+void ChannelAttachmentsBase::Erase(const std::shared_ptr<Track> &parent,
+   size_t index)
 {
-   if (index < mAttachments.size())
+   auto size = mAttachments.size();
+   if (index < size) {
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-track/ChannelAttachments.cpp:160]: It is intentionally not stated as a precondition.

The idea of reparenting is that each attachment is informed when its numerical slot in the array changes.  There is no change for those before `index`.

File: libraries/lib-wave-track/WaveTrack.cpp (hunk 11 above)
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.cpp:4272]: Precisely not!

The tentative stereo channel pair (as indicated by the "leader" attribute) is discovered to be ineligible to merge and so two mono tracks remain.

File: libraries/lib-wave-track/CMakeLists.txt (hunk 12 above)
Comment (Paul-Licameli) [reply libraries/lib-wave-track/CMakeLists.txt:38]: Yes, it simplified the job, avoiding even more const-casting in WaveTrack.cpp

**Review by saintmatthieu — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
Merged master to wave-clip-refactoring, then rebased this with some squashing, and added to the QA checklist

**Comment by Paul-Licameli** [comment]
I discovered a problem, maybe noticeable only in debug builds.  Generate or record some sound, then use Reverse effect.  An assertion is violated.

I also observe that tracks can have duplicate shared pointers to a clip.

**Comment by saintmatthieu** [comment]
> I discovered a problem, maybe noticeable only in debug builds. Generate or record some sound, then use Reverse effect. An assertion is violated.
> 
> I also observe that tracks can have duplicate shared pointers to a clip.

@chinakov should be aware - he currently is actively testing this card.

**Comment by Paul-Licameli** [comment]
@saintmatthieu please look at the two newest commits
@chinakov I'm afraid you might need to retest it all to be very sure

File: libraries/lib-wave-track/WaveTrack.cpp [hunk 14]
```diff
@@ -4240,10 +4241,14 @@ bool WaveChannel::ClipsAreUnique() const
 {
    // This is used only in assertions
    auto intervals = Intervals();
-   const auto begin = intervals.begin();
```
Comment (saintmatthieu) [inline libraries/lib-wave-track/WaveTrack.cpp]: Ouch, of course ...

**Review by saintmatthieu — APPROVED** [review]

**Pull request merged**